components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/WebsocketHandler.java

```diff
@@ -0,0 +1,437 @@
+			String apikey = auth[1];
+			if (util.isRemoveOAuthHeadersFromOutMessage()) {
+				req.headers().remove(HttpHeaders.AUTHORIZATION);
+			}
+			//If the key have already been validated
+			if (util.isGatewayTokenCacheEnabled()) {
+				info = util.validateCache(apikey, apikey);
+				if (info != null) {
+					if (info.isAuthorized()) {
+						return validateSubscriptionDetails(api.getContext(),
+						                                   api.getId().getVersion(),
+						                                   info.getConsumerKey());
+					} else
+						return false;
+				}
+			}
+			try {
+				tokenInfo = KeyManagerHolder.getKeyManagerInstance().getTokenMetaData(apikey);
+				if (tokenInfo == null) {
+					return false;
+				}
+			} catch (APIManagementException e) {
+
+				e.printStackTrace();
+			}
+			if (util.isGatewayTokenCacheEnabled()) {
+				info = new APIKeyValidationInfoDTO();
+				info.setAuthorized(tokenInfo.isTokenValid());
+				info.setEndUserName(tokenInfo.getEndUserName());
+				info.setConsumerKey(tokenInfo.getConsumerKey());
+				info.setIssuedTime(tokenInfo.getIssuedTime());
+				info.setValidityPeriod(tokenInfo.getValidityPeriod());
+				util.putCache(info, apikey, apikey);
+			}
+			//apiSubscriber = apiConsumer.getSubscriberById(apikey);
+
+			if (tokenInfo.isTokenValid()) {
+
+				return validateSubscriptionDetails(api.getContext(), api.getId().getVersion(),
+				                                   tokenInfo.getConsumerKey());
+
+			} else {
+				return false;
+			}
+
+		} else {
+			return false;
+		}
+	}
+
+	private boolean throttle2(ChannelHandlerContext ctx) throws APIManagementException {
+
+		AuthUtil util = new AuthUtil();
+		if (throttleDataPublisher == null) {
+			// The publisher initializes in the first request only
+			synchronized (this) {
+				throttleDataPublisher = new ThrottleDataPublisher();
+			}
+		}
+
+		String applicationLevelTier = infoDTO.getApplicationTier();
+		String apiLevelTier = infoDTO.getApiTier();
+		String subscriptionLevelTier = infoDTO.getTier();
+		String resourceLevelTier = apiLevelTier;
+		String authorizedUser = infoDTO.getSubscriber() + "@" + infoDTO.getSubscriberTenantDomain();
+		String apiName = infoDTO.getApiName();
+		String apiContext = api.getContext();
+		String apiVersion = api.getId().getVersion();
+		String appTenant = infoDTO.getSubscriberTenantDomain();
+		String apiTenant = tenantDomain;
+		String appId = infoDTO.getApplicationId();
+		String applicationLevelThrottleKey = appId + ":" + authorizedUser;
+		String apiLevelThrottleKey = apiContext + ":" + apiVersion;
+		String resourceLevelThrottleKey = apiLevelThrottleKey;
+		String subscriptionLevelThrottleKey = appId + ":" + apiContext + ":" + apiVersion;
+		String messageId = UIDGenerator.generateURNString();
+		String remoteIP = ctx.channel().remoteAddress().toString();
+		remoteIP = remoteIP.substring(1, remoteIP.indexOf(":"));
+		JSONObject jsonObMap = new JSONObject();
+		if (remoteIP != null && remoteIP.length() > 0) {
```

Comment:
remoteIP.length() > 0 should be flipped in order to avoid null pointer exceptions